### PR TITLE
DB hard reset

### PR DESCRIPTION
**Pull request by MarinPostma**
This PR implements DB hardreset. This allows restarting a replica that is in an invalid state, wiping it's database clean so it can start replicating from the primary from scratch.

### CODE REVIEW

File: sqld/src/lib.rs [hunk 1]
```diff
@@ -5,15 +5,19 @@ use std::sync::Arc;
 #[cfg(feature = "mwal_backend")]
 use std::sync::Mutex;
 use std::task::{Context, Poll};
+use std::time::Duration;
```
Comment (haaawk) [inline sqld/src/lib.rs]: Github says it's unused.

File: sqld/src/lib.rs [hunk 2]
```diff
@@ -99,9 +112,20 @@ where
...
 
-    while let Some(res) = handles.next().await {
-        res??;
-    }
+    Ok(())
+}
+
+/// nukes current DB and start anew
+async fn hard_reset(config: &Config, mut handles: Handles) -> anyhow::Result<()> {
+    tracing::error!("received hard-reset command: reseting replica.");
```
Comment (haaawk) [inline sqld/src/lib.rs:126]: Probably warning is better. This is not really an error.

File: sqld/src/lib.rs [hunk 3]
```diff
@@ -99,9 +112,20 @@ where
...
+
+    tracing::info!("Shutting down all services...");
+    handles.iter_mut().for_each(|h| h.abort());
+    while let Some(_) = handles.next().await {}
```
Comment (haaawk) [inline sqld/src/lib.rs]: We don't want to report potential errors returned from handles?

File: sqld/src/lib.rs [hunk 4]
```diff
@@ -144,87 +225,29 @@ pub async fn run_server(config: Config) -> anyhow::Result<()> {
...
-                #[cfg(feature = "mwal_backend")]
-                vwal_methods,
-            )
-            .await
-            .context("failed to start WriteProxy DB")?;
-            let service = DbFactoryService::new(factory);
-            run_service(service, config).await?;
+        match config.writer_rpc_addr {
+            Some(ref addr) => start_primary(&config, &mut handles, addr).await?,
+            None => start_replica(&config, &mut handles).await?,
```
Comment (haaawk) [inline sqld/src/lib.rs:234]: Putting extraction of start_primary and start_replica on separate commits would make it easier to review.

Comment (haaawk) [reply sqld/src/lib.rs:234]: Just a tip for the future.

File: sqld/src/lib.rs [hunk 5]
```diff
@@ -144,87 +225,29 @@ pub async fn run_server(config: Config) -> anyhow::Result<()> {
...
-                config.db_path.clone(),
-                #[cfg(feature = "mwal_backend")]
-                vwal_methods,
-            )
-            .await
-            .context("failed to start WriteProxy DB")?;
-            let service = DbFactoryService::new(factory);
-            run_service(service, config).await?;
+        match config.writer_rpc_addr {
+            Some(ref addr) => start_primary(&config, &mut handles, addr).await?,
```
Comment (haaawk) [inline sqld/src/lib.rs:233]: wouldn't it be cleaner for start_primary and start_replica to return a handle that this function adds to handles?

File: sqld/src/lib.rs (hunk 3 above)
Comment (MarinPostma) [reply sqld/src/lib.rs]: the error is always going to be an abort

File: sqld/src/lib.rs [hunk 6]
```diff
@@ -42,6 +46,16 @@ pub enum Backend {
...
+/// Services handles registry.
+/// All created services must be registered here, so they can be awaited together.
+type Handles = FuturesUnordered<JoinHandle<anyhow::Result<()>>>;
+
+/// Trigger a hard database reset. This cause the database to be wiped, freshly restarted
+/// This is used for replicas that are left in an unrecoverabe state and should restart from a
+/// fresh state.
+///
+/// /!\ use with caution.
+pub(crate) static HARD_RESET: Lazy<Arc<Notify>> = Lazy::new(|| Arc::new(Notify::new()));
```
Comment (haaawk) [inline sqld/src/lib.rs:59]: Does this have to be a global?

File: sqld/src/lib.rs (hunk 5 above)
Comment (haaawk) [reply sqld/src/lib.rs:233]: Or collection of handles if there's more than one.

File: sqld/src/database/libsql.rs [hunk 7]
```diff
@@ -1,7 +1,5 @@
 use std::path::Path;
 use std::str::FromStr;
-#[cfg(feature = "mwal_backend")]
```
Comment (haaawk) [inline sqld/src/database/libsql.rs:3]: @psarna Could you please review this commit?

**Review by haaawk — COMMENTED** [review]
LGTM but I think it would be valuable for @psarna to have a look at mwal changes

**Comment by MarinPostma** [comment]
bors merge

**Pull request merged**